wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,709 @@
+
+# Values from CumlArrayDescriptorMeta can be specifically read
+print(my_est.__dict__["my_cuml_array_"].input_type)
+# Output: "cupy"
+
+# The input value can be accessed
+print(my_est.__dict__["my_cuml_array_"].get_input_value())
+# Output: CumlArray ...
+```
+
+#### Internal Array-Like Conversion
+
+Internally, developers can use the same `global_output_type` functionality to convert arrays as well. This has the same benefits of lazy conversion and caching as when descriptors are used externally. For example, we could add a `score()` function to `TestEstimator`:
+
+```python
+def score(self):
+
+   # Set the global output type to numpy
+   with cuml.using_output_type("numpy"):
+      # Accessing my_cuml_array_ will return a numpy array and
+      # the result can be returned directly
+      return np.sum(self.my_cuml_array_, axis=0)
+```
+
+### Estimator Functions
+
+In order for estimator functions to accept a wide variety of array types and correctly return the user's desired type, a fair amount of boiler plate code is involved. These common steps were often repetitive, inefficient, and incorrectly implemented.
+
+To assit with this, a set of decorators have been created to wrap estimator functions (and all `cuml` API functions as well) and perform the boiler plate actions automatically.
+
+#### Common Functionality
+
+All of these decorators perform the same basic steps with a few small differences. The common steps performed by each decorator is:
+
+1. Set `cuml.global_output_type = "mirror"`
+   1. When `"mirror"` is used as the global output type, that indicates we are in an internal cuML API call. The `CumlArrayDescriptor` keys off this value to change between internal and external functionality
+2. Set CuPy allocator to use RMM
+   1. This replaces the existing decorator `@with_cupy_rmm`
+   2. Unlike before, the CuPy allocator is only set once per API call
+3. Set the estimator input attributes. Can be broken down into 3 steps:
+   1. Set `_input_type` attribute
+   2. Set `target_dtype` attribute
+   3. Set `n_features` attribute
+4. Call the desired function
+5. Get the estimator output type. Can be broken down into 2 steps:
+   1. Get `output_type`
+   2. Get `output_dtype`
+6. Convert the return value
+   1. This will ultimately call `CumlArray.to_output(output_type=output_type, output_dtype=output_dtype)
+
+While the above list of steps may seem excessive for every call, most functions follow this general form, but may skip a few steps depending on a couple of factors. For example, Step #3 is necessary on functions that modify the estimator's estimated attributes, such as `fit()`, but is not necessary for functions like `predict()` or `transform()`. And Step #5/6 are only necessary when returning array-like objects and are ommitted when returning any other type.
+
+Functionally, you can think of these decorators equivalent to the following pseudocode:
```

Comment:
Yeah, I would definitely put this after showing how decorators are used and call it details or internals

Reply:
Will do

Reply:
Done. Resolving comment